### ISSUE TITLE
BFGS could accept an initial inverse hessian like TF

### ISSUE THREAD

**Issue by davidia** [google/jax#5139]
#4594 could be solved and performance improved if it were possible to supply an the initial state of the hessian or it's inverse as in the [TF probability implementation](https://www.tensorflow.org/probability/api_docs/python/tfp/optimizer/bfgs_minimize).

Would you consider a PR for this?

**Comment by shoyer** [comment]
I'm open to this, but I guess the natural question is why this feature makes sense for JAX if it doesn't exist in SciPy?

**Comment by Joshuaalbert** [comment]
@davidia I experimented with this when implementing the BFGS for JAX and it's not a good idea in many situations. Firstly, not all functions have well-defined Hessian inverses. Secondly, even when there is, this can lead to poor BFGS convergence behaviour. I determined this empirically. At the moment, I think this feature should not be added for stability reasons and because it would deviate from SciPy. However, I'd be open to exploring this more.

**Comment by Joshuaalbert** [comment]
#4594 is already solved by #4594. It's just waiting to be merged @shoyer can you poke it?

**Comment by LoicRaillon** [comment]
I tried the Jax implementation of the BFGS but the algorithm doesn't find a proper line search at the first iteration whereas the Scipy implementation works very well. By digging in both implementations, it seems that the inverse Hessian is initialized with the identity matrix but  Nocedal and Wright in "Numerical Optimization p.143 formula (6.20)" proposed an heuristic method to scale the Hessian matrix at the first iteration. This is implemented in Scipy [FullHessianUpdateStrategy](https://github.com/scipy/scipy/blob/v1.5.4/scipy/optimize/_hessian_update_strategy.py#L146-L158) but it seems is not used by their BFGS algorithm. What do you think? Is it worth testing?

**Comment by davidia** [comment]
@Joshuaalbert got it, thanks.

**Comment by Joshuaalbert** [comment]
@LoicRaillon Can you post a minimal example showing this behaviour? Also, keep in mind line-search behaviour isn't necessarily related to the inverse Hessian. Line search only requires a search direction and then it approximately minimises a scalar function along that line within some constraints. 

The BFGS as it exists in scipy, and here in JAX starts with an initial inverse Hessian of identity. While there are other options, often the identity is a good enough starting point (this is also said by Nocedal and Wright). The reasoning is that the extra costs typically don't improve the solution. Remember BFGS and it's variants far variable metric methods, and start of as gradient descent and asymptote to approximate Newton methods. The speed of convergence is often fast (only a few iterations). However, I am interested in exploring this a bit more. Therefore, I would need some examples.
